### PR TITLE
OCLOMRS-482: Fix Mapping relationship in Edit and Create to allow cha…

### PR DESCRIPTION
**Pull request by myrdstom**
# JIRA TICKET NAME:
[Fix Mapping relationship in Edit and Create to allow changing to other relationship types.](https://issues.openmrs.org/browse/OCLOMRS-482)

# Summary:

- When creating or editing a new mapping, if a user changed the source from CIEL to anything else, the user could not edit the relationship field from the default value: Narrower-than. 

- Currently, when a user changes the value for the source from CIEL to anything else, he should be able to change the value for the relationship to any field.

### CODE REVIEW

File: src/components/dictionaryConcepts/components/MapType.jsx [hunk 1]
```diff
@@ -24,6 +25,7 @@ const mapType = (props) => {
         {' '}
       </option>)
     }
+    console.log(props);
```
Comment (danuluma) [inline src/components/dictionaryConcepts/components/MapType.jsx]: Please remove the console.log statement

File: src/components/dictionaryConcepts/components/CreateMapping.jsx [hunk 2]
```diff
@@ -2,26 +2,55 @@ import React, { Component } from 'react';
...
+  }
+
+  componentDidUpdate(prevProps) {
+    if (prevProps.source !== this.props.source) {
+      // eslint-disable-next-line react/no-did-update-set-state
+      this.setState({
+        type: '',
+      });
+    }
+    console.log('Hit CDU');
```
Comment (danuluma) [inline src/components/dictionaryConcepts/components/CreateMapping.jsx]: Kindly remove the console.log statemnet

Comment (myrdstom) [reply src/components/dictionaryConcepts/components/CreateMapping.jsx]: The console.log statements are there purely for testing purposes. When I bulk up my test coverage, I will remove them

File: env-config.js [hunk 3]
```diff
@@ -0,0 +1,2 @@
+window.env = {
```
Comment (dkayiwa) [inline env-config.js]: How is this related to the ticket?

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 4]
```diff
@@ -74,4 +74,18 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+  it('should call componentDidUpdate', () =>{
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Are you testing a built in reactjs method? 
I do not see the relevance of this ticket in relation to what the ticket is supposed to fix.

Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have renamed it to better explain what it does

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 5]
```diff
@@ -74,4 +74,18 @@ describe('Test suite for dictionary concepts components', () => {
...
+      isNew: true,
+      source: 'Source',
+    };
+    wrapper = mount(<CreateMapping {...newProps} />);
+    const nextProps = {
+      ...props,
+      source: 'MapType',
+    };
+    wrapper.setProps(nextProps);
+    expect(wrapper.instance().props.source).toEqual('MapType');
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: How is this related to "should set the state to an empty string...."?

Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have renamed it appropriately

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 6]
```diff
@@ -74,4 +74,18 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+  it('should adjust the props for source', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Are you testing functionality of the ReactJS library? Or are you testing for a certain OCL use case?

Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: If I follow you correctly. You want an explanation of what happens when the test runs as per OCL. I have implemented that

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 7]
```diff
@@ -74,4 +75,18 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+  it('should render the new source to equal MapType', () => {
+    const newProps = {
+      ...props,
+      isNew: true,
+      source: 'CIEL',
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Don't we already have a constant for this?

Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: yes, we do... I'll adjust accordingly

Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Done

**Comment by dkayiwa** [comment]
What does a commit message of "mend" mean?

**Comment by myrdstom** [comment]
> What does a commit message of "mend" mean?

I must have mistyped while writing a commit message, I have resolved it

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 8]
```diff
@@ -74,4 +75,18 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+  it('should render the new source to equal MapType', () => {
+    const newProps = {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: How is this test related to the ticket description?

Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: In the previous implementation, the reason 'relationships' data could not be changed  is because if the fields had been hardcoded in that if source was equal to CIEL, then the relationship was equal to 'SAME-AS' and if source was not equal to 'CIEL', the relationship was equal to 'NARROWER THAN'. As such, I did not have access to the 'relationship' field

One of the first steps in implementing this task was making sure that the source field in relation to the relationship field was not hardcoded but more dynamic and that is what I did with this test testing the dynamic change in the source field.

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 9]
```diff
@@ -74,4 +75,13 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+  it('should render the new source to equal MapType', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: How does this test ensure that a user can edit the 'relationship' field appropriately?

**Comment by dkayiwa** [comment]
Can you also take a look at this? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js (hunk 9 above)
Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have refactored the test.  Is it clearer now? Otherwise, please elaborate on what I'm missing.

**Comment by dkayiwa** [comment]
Can you take a second look at this? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by myrdstom** [comment]
> Can you take a second look at this? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

I looked through it and I adjusted the description of the PR and the ticket. I believe it is okay for you to review

**Comment by dkayiwa** [comment]
Can you ask your team mates to take a looks at the pull request tips page in reference to your pull request?

**Comment by myrdstom** [comment]
> Can you ask your team mates to take a looks at the pull request tips page in reference to your pull request?

I have looked through this PR

**Comment by myrdstom** [comment]
I have rebased and edited the description of the PR. Please confirm that the necessary changes have been made.

**Comment by myrdstom** [comment]
have adjusted my commit message as per standard. Please confirm that the necessary changes have been made.

File: src/tests/dictionaryConcepts/components/CreateMapping.test.js [hunk 10]
```diff
@@ -74,4 +75,14 @@ describe('Test suite for dictionary concepts components', () => {
     const inputs = wrapper.find('select');
     expect(inputs).toHaveLength(2);
   });
+  it('should render the new relationship which is equal to Associated With', () => {
+    wrapper = mount(<CreateMapping {...props} />);
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Can you describe how this ticket is related to the ticket description?

Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: The test is checking that when that the source is changed the relationship can also be adjusted.  It is related to my ticket because it is what I was fixing. As per the ticket description. When a user changed the source from CIEL to anything else, the user could not adjust the relationship field.

Comment (dkayiwa) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: Can the description explicitly state what should lead to the rendering of the new relationship?

Comment (myrdstom) [reply src/tests/dictionaryConcepts/components/CreateMapping.test.js]: I have updated the description. Is it explicit enough?
